src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java

```diff
@@ -0,0 +1,140 @@
+        int[] turnAroundTime = new int[noOfProcesses];
+        boolean[] finishedProcess = new boolean[noOfProcesses];
+        int finishedProcessCount = 0;
+
+        // Sort by arrival times using a simple bubble sort for demonstration
+        int[] sortedIndices = sortIndicesByArrival(arrivalTimes);
+        arrivalTimes = Arrays.copyOf(arrivalTimes, arrivalTimes.length);
+        burstTimes = Arrays.copyOf(burstTimes, burstTimes.length);
+        processNames = Arrays.copyOf(processNames, processNames.length);
+
+        // Reorder the arrays based on sorted indices
+        int[] tempArrival = new int[noOfProcesses];
+        int[] tempBurst = new int[noOfProcesses];
+        String[] tempProcess = new String[noOfProcesses];
+
+        for (int i = 0; i < noOfProcesses; i++) {
+            tempArrival[i] = arrivalTimes[sortedIndices[i]];
+            tempBurst[i] = burstTimes[sortedIndices[i]];
+            tempProcess[i] = processNames[sortedIndices[i]];
+        }
+
+        arrivalTimes = tempArrival;
+        burstTimes = tempBurst;
+        processNames = tempProcess;
+
+        while (finishedProcessCount < noOfProcesses) {
+            currentTime = Math.max(currentTime, arrivalTimes[findNextProcess(arrivalTimes, finishedProcess, currentTime)]);
+            int loc = findHighestResponseRatio(processNames, arrivalTimes, burstTimes, finishedProcess, currentTime);
+
+            turnAroundTime[loc] = currentTime + burstTimes[loc] - arrivalTimes[loc];
+            currentTime += burstTimes[loc];
+            finishedProcess[loc] = true;
+            finishedProcessCount++;
+        }
+
+        return turnAroundTime;
+    }
+
+    /**
+     * Calculates the Waiting Time (WT) for each process.
+     *
+     * @param turnAroundTime The Turn Around Times for each process.
+     * @param burstTimes     The burst times for each process.
+     * @return An array of Waiting Times for each process.
+     */
+    public static int[] calculateWaitingTime(int[] turnAroundTime, int[] burstTimes) {
+        int[] waitingTime = new int[turnAroundTime.length];
+        for (int i = 0; i < turnAroundTime.length; i++) {
+            waitingTime[i] = turnAroundTime[i] - burstTimes[i];
+        }
+        return waitingTime;
+    }
+
+    /**
+     * Finds the next process to be scheduled based on arrival times and the current time.
+     *
+     * @param arrivalTimes    Array of arrival times for each process.
+     * @param finishedProcess Array indicating whether each process has finished.
+     * @param currentTime     The current time in the scheduling process.
+     * @return The index of the next process to be scheduled.
+     */
+    private static int findNextProcess(int[] arrivalTimes, boolean[] finishedProcess, int currentTime) {
+        for (int i = 0; i < arrivalTimes.length; i++) {
+            if (!finishedProcess[i] && arrivalTimes[i] <= currentTime) {
+                return i;
+            }
+        }
+        return 0;
```

Comment:
Can this value, cause problems if the process with index 0 is not ready yet?

If yes, for such case better instead current approach, to return -1 or another special index to explicitly indicate that the process was not found. In the loop body, it is worth checking that the process actually exists.

Reply:
I have changed this to the following:

```java
int maxIndex = PROCESS_NOT_FOUND; // PROCESS_NOT_FOUND = -1
...
// Find the process with the highest response ratio
...
return maxIndex;
```